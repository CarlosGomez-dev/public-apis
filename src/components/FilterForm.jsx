import { useEffect, useState } from 'react';
import { useFetchApiCategories } from '../hooks/useFetchApiCategories';
import { FilterFormSelect } from './FilterFormSelect';

const PAGE_SIZE_OPTIONS = [10, 20, 50, 100];
const apiKeysAsOptions = [
  ['API', 'Name'],
  'Description',
  'Auth',
  'HTTPS',
  'Cors',
  'Link',
  'Category',
];

export const FilterForm = ({
  handleSearch,
  pageLimit,
  setPageLimit,
  setSortBy,
  handleCategory,
}) => {
  const [inputText, setInputText] = useState('');
  const categories = useFetchApiCategories();

  const handleInput = event => {
    setInputText(event.target.value);
  };

  useEffect(() => {
    const timer = setTimeout(() => handleSearch(inputText), 500);
    return () => clearTimeout(timer);
  }, [handleSearch, inputText]);

  return (
    <form onSubmit={e => e.preventDefault()}>
      <label htmlFor='filter'>Find API:</label>
      <input type='text' id='filter' value={inputText} onChange={handleInput} />
      {/* <button type='submit'>Search</button> */}

      <FilterFormSelect
        label='Results per Page'
        optionsArray={PAGE_SIZE_OPTIONS}
        defaultValue={pageLimit}
        onChange={setPageLimit}
      />

      <FilterFormSelect
        label='Sort by'
        optionsArray={apiKeysAsOptions}
        onChange={setSortBy}
        hasNullOption
      />

      <FilterFormSelect
        label='Categories'
        optionsArray={categories}
        onChange={handleCategory}
        hasNullOption
      />
    </form>
  );
};
