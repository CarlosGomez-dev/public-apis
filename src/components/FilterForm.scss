@import '../colors';

.form {
  display: grid;
  padding: 0.5rem;
  span {
    font-size: 0.7rem;
    padding: 0.3rem 0.5rem;
    margin-bottom: 1rem;
  }
}

.label {
  text-transform: uppercase;
  font-size: 0.75rem;
}

.input,
.select {
  border: 1px solid $border-color;
  border-radius: 5px;
  padding: 0.5rem 1rem;
}

.search {
  padding: 0.7rem 1.2rem;
  font-size: 1rem;
  line-height: 1rem;
}

.form-filters {
  display: flex;
  flex-wrap: wrap;
  gap: 0.5rem;
}

.form-select {
  display: grid;
  flex-grow: 1;
}

@media (min-width: 768px) {
  .form-filters {
    display: grid;
    grid-template-columns: auto auto;
  }
  .sort-by-select {
    display: none;
  }
}
