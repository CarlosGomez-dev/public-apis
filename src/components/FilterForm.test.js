import { act, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { FilterForm } from './FilterForm';

const handleSearch = jest.fn();
const filterFormProps = { handleSearch };

describe('FilterForm', () => {
  it('matches snapshot', () => {
    const { container } = render(<FilterForm {...filterFormProps} />);
    expect(container).toMatchSnapshot();
  });
  it('accepts user input', () => {
    render(<FilterForm {...filterFormProps} />);
    const inputText = 'weather';
    const input = screen.getByLabelText(/search/i);
    jest.useFakeTimers();
    userEvent.type(input, inputText + '{enter}');
    expect(setTimeout).toHaveBeenCalled();
    act(() => jest.runAllTimers());
    expect(handleSearch).toHaveBeenCalled();
  });
});
