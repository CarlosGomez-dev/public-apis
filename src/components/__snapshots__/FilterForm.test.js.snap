// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterForm matches snapshot 1`] = `
<div>
  <form
    class="form"
  >
    <label
      class="label"
      for="filter"
    >
      Find
    </label>
    <input
      class="input search"
      id="filter"
      placeholder="e.g.: Weather"
      type="text"
      value=""
    />
    <span>
      Press / to jump to the search box
    </span>
    <section
      class="form-filters"
    >
      <div
        class="form-select "
      >
        <label
          class="label"
          for="Results-per-Page"
        >
          Results per Page
        </label>
        <select
          class="select"
          id="Results-per-Page"
        >
          <option
            value="10"
          >
            10 / page
          </option>
          <option
            value="20"
          >
            20 / page
          </option>
          <option
            value="50"
          >
            50 / page
          </option>
          <option
            value="100"
          >
            100 / page
          </option>
        </select>
      </div>
      <div
        class="form-select sort-by-select"
      >
        <label
          class="label"
          for="Sort-by"
        >
          Sort by
        </label>
        <select
          class="select"
          id="Sort-by"
        >
          <option
            value=""
          >
            -
          </option>
          <option
            value="API"
          >
            Name
          </option>
          <option
            value="Description"
          >
            Description
          </option>
          <option
            value="Auth"
          >
            Auth
          </option>
          <option
            value="HTTPS"
          >
            HTTPS
          </option>
          <option
            value="Category"
          >
            Category
          </option>
        </select>
      </div>
      <div
        class="form-select "
      >
        <label
          class="label"
          for="Categories"
        >
          Categories
        </label>
        <select
          class="select"
          id="Categories"
        >
          <option
            value=""
          >
            -
          </option>
        </select>
      </div>
    </section>
  </form>
</div>
`;
